pkg/chunk/elastic/es_index_client.go

```diff
@@ -0,0 +1,344 @@
+
+	case len(query.RangeValuePrefix) > 0 && query.ValueEqual != nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValuePrefix)).
+			Lt(string(query.RangeValuePrefix) + null)
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case len(query.RangeValueStart) > 0 && query.ValueEqual == nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValueStart))
+
+	case len(query.RangeValueStart) > 0 && query.ValueEqual != nil:
+		rangeQuery = elastic.NewRangeQuery("range").Gte(string(query.RangeValueStart))
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case query.ValueEqual != nil:
+		valueTermQuery = elastic.NewTermQuery("value", query.ValueEqual)
+
+	case query.ValueEqual == nil:
+		break
+	}
+
+	exists, err := e.client.IndexExists(query.TableName).Do(ctx)
+	if err != nil {
+		// Handle error
+		fmt.Println(exists)
+	}
+
+	// Search with a term query
+	baseQuery := e.client.Search().
+		Index(query.TableName).
+		Query(hashTermQuery)
+	if valueTermQuery != nil {
+		baseQuery = baseQuery.Query(valueTermQuery)
+	}
+	if rangeQuery != nil {
+		baseQuery = baseQuery.Query(rangeQuery)
+	}
+
+	// Search with a term query
+	searchResult, err := baseQuery.
+		Sort("range", true). // sort by "range" field, ascending
+		From(0).Size(maxFetchDocs).
+		Pretty(true). // pretty print request and response JSON
+		Do(ctx)       // execute
+
+	if searchResult == nil || searchResult.Hits == nil {
+		return nil
+	}
+
+	if err != nil {
+		// Handle error
+		level.Error(util.Logger).Log("msg", fmt.Sprintf("Query in index %s met error!", query.TableName))
+		panic(err)
+	}
+
+	// searchResult is of type SearchResult and returns hits, suggestions,
+	// and all kinds of other information from Elasticsearch.
+	level.Debug(util.Logger).Log("msg", fmt.Sprintf("Query took %d milliseconds with result num\n", searchResult.TookInMillis))
+
+	var batch readBatch
+	var ttyp IndexEntry
+	for _, item := range searchResult.Each(reflect.TypeOf(ttyp)) {
+		if t, ok := item.(IndexEntry); ok {
+			level.Debug(util.Logger).Log("msg", fmt.Sprintf("Index by hash %s: range %s, value %s\n", t.Hash, t.Range, t.Value))
+			batch.rangeValue = []byte(t.Range)
+			batch.value = []byte(t.Value)
+
+			if !callback(&batch) {
+				return nil
+			}
+		}
+	}
+
+	return nil
+}
+
+// NewESIndexClient creates a new IndexClient that used ElasticSearch.
+func NewESIndexClient(cfg Config) (chunk.IndexClient, error) {
+	client, err := newES(cfg)
+	if err != nil {
+		return nil, err
+	}
+	indexClient := &esClient{
+		cfg,
+		client,
+	}
+	return indexClient, nil
+}
+
+func newES(cfg Config) (*elastic.Client, error) {
+	//fix x509: certificate signed by unknown authority
+	var tr *http.Transport
+	if cfg.TLSSkipVerify { // if skip TLS Verify
+		tr = &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+	} else { // not skip TLS Verify
+		cert, err := tls.LoadX509KeyPair(cfg.CertFile, cfg.KeyFile)
```

Comment:
no option to use with valid certificates that are already in rootcabundle? Now this assumes 1) insecure 2) certs provided in start flags.

Reply:
I haven't used the way you mentioned. Could you share some details?